Purge old ip on change

gusev-p opened the pull request:
When a node changes IP address we need to remove its old IP from `system.peers` and gossiper.

We do this in `sync_raft_topology_nodes` when the new IP is saved into `system.peers` to avoid losing the mapping if the node crashes between deleting and saving the new IP. We also handle the possible duplicates in this case by dropping them on the read path when the node is restarted.

The PR also fixes the problem with old IPs getting resurrected when a node changes its IP address.
The following scenario is possible: a node `A` changes its IP from `ip1` to `ip2` with restart, other nodes are not yet aware of `ip2` so they keep gossiping `ip1`. After restart `A` receives `ip1` in a gossip message and calls `handle_major_state_change` since it considers it as a new node. Then `on_join` event is called on the gossiper notification handlers, we receive such event in `raft_ip_address_updater` and reverts the IP of the node A back to ip1.

To fix this we ensure that the new gossiper generation number is used when a node registers its IP address in `raft_address_map` at startup.

The `test_change_ip` is adjusted to ensure that the old IPs are properly removed in all cases, even if the node crashes.

Fixes #16886
Fixes #16691
Fixes #17199

scylladb-promoter commented:
#### :red_circle: CI State: FAILURE

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/6405/)
:white_check_mark: - [Container Test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/6405/)
:white_check_mark: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/7655/)
:x: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/6405/)
#### Failed Tests (6/24309):
- [test_change_two](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/6405/testReport/junit/%28root%29/test_change_ip/Tests___Unit_Tests___test_change_two) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_change_two)
- [test_change_two](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/6405/testReport/junit/%28root%29/test_change_ip/Tests___Unit_Tests___test_change_two_2) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_change_two)
- [test_change_two](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/6405/testReport/junit/%28root%29/test_change_ip/Tests___Unit_Tests___test_change_two_3) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_change_two)
- [topology.test_change_ip.release.1](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/6405/testReport/junit/%28root%29/non-boost%20tests/Tests___Unit_Tests___topology_test_change_ip_release_1) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+topology.test_change_ip.release.1)
- [topology.test_change_ip.release.2](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/6405/testReport/junit/%28root%29/non-boost%20tests/Tests___Unit_Tests___topology_test_change_ip_release_2) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+topology.test_change_ip.release.2)
- [topology.test_change_ip.release.3](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/6405/testReport/junit/%28root%29/non-boost%20tests/Tests___Unit_Tests___topology_test_change_ip_release_3) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+topology.test_change_ip.release.3)
#### Build Details:
- Duration: 1 hr 53 min
- Builder: spider1.cloudius-systems.com

gusev-p commented:
new version:
  * turn off the parts of `test_change_two` that rely on error injection in release mode.

[changes](https://github.com/scylladb/scylladb/compare/f733b97454adc2589cbe23cf43257ac3a198ea69..b20b66fe92893de05941b194d3f69ff31bc3822a)

scylladb-promoter commented:
#### :green_circle: CI State: SUCCESS

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/6414/)
:white_check_mark: - [Container Test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/6414/)
:white_check_mark: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/7664/)
:white_check_mark: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/6414/)
#### Build Details:
- Duration: 2 hr 34 min
- Builder: i-012791b0677ae7763 (m5d.12xlarge)

kbr-scylla commented on service/storage_service.cc:
```diff
@@ -345,7 +345,10 @@ static locator::node::state to_topology_node_state(node_state ns) {
 
 // Synchronizes the local node state (token_metadata, system.peers/system.local tables,
 // gossiper) to align it with the other raft topology nodes.
-future<> storage_service::sync_raft_topology_nodes(mutable_token_metadata_ptr tmptr, std::optional<locator::host_id> target_node, std::unordered_set<raft::server_id> prev_normal) {
```
re commit message / cover letter: you mean `system.peers`

gusev-p replied:
fixed

kbr-scylla commented on service/storage_service.cc:
```diff
@@ -683,23 +683,23 @@ class storage_service::raft_ip_address_updater: public gms::i_endpoint_state_cha
 
     future<>
-    on_endpoint_change(gms::inet_address endpoint, gms::endpoint_state_ptr ep_state) {
+    on_endpoint_change(gms::inet_address endpoint, gms::endpoint_state_ptr ep_state, const char* ev) {
```
Why the abbreviation though. How about `event_name`?

kbr-scylla commented on db/system_keyspace.cc:
```diff
@@ -1464,19 +1464,17 @@ struct local_cache {
         }
-        gms::inet_address gms_addr(std::move(peer));
-        sstring dc = row.template get_as<sstring>("data_center");
-        sstring rack = row.template get_as<sstring>("rack");
```
Some people would argue that introducing such variables makes the code more readable. I don't have a strong opinion, but it's preferred to avoid such subjective stylistic changes. (I won't oppose this change.)

(The coroutinization below is definitely desired though.)

kbr-scylla commented on db/system_keyspace.cc:
```diff
@@ -1463,7 +1463,47 @@ struct local_cache {
     system_keyspace::bootstrap_state _state;
 };
```
> The new function also removes the records with no host_id, so we no longer need the get_host_id function.

I'm not sure we want to do that. It could mask a bug -- such records should never appear in the first place.

gusev-p replied:
> Could also be that gossiper was missing HOST ID application state due to a bug (and then we persisted such entry in system.peers) -- and recently we're seeing errors for missing HOST ID, I think they only appear for removed nodes, but it's still unclear why, I believe they should never appear, a node either should have host id attached or not appear at all.

We can check if `host_id` is empty in `system_keyspace::update_peer_info` and trigger `on_internal_error` if it is. I don't know why we don't have such check yet.

My argument was that both of these inconsistencies can occur not only due to node crashes but also due to bugs, so why handle them differently then.

kbr-scylla replied:
> We can check if host_id is empty in system_keyspace::update_peer_info and trigger on_internal_error if it is. I don't know why we don't have such check yet.

Makes sense. @bhalevy please also give your opinion as you were modifying this code recently and probably thought about the problem too

bhalevy replied:
It was supposed to be covered in https://github.com/scylladb/scylladb/pull/16376.
There could be use cases to update other fields in system.peers (well the host_id should never change - it really should be the partition key).  So we can require it in `update_peer_info` and do read before write to make sure it's only set once, and in all other cases we can verify it matches the one given by the caller.

kbr-scylla replied:
> well the host_id should never change - it really should be the partition key

It does change -- replace-with-same-IP

bhalevy replied:
Yeah, well, it's not the same node anymore.
It's a replacing node that assumes ownership of the same tokens.
But I agree that makes the verification method I suggest impractical, unless we care to delete the previous record (a-la removenode) prior to inserting of the new one.

gusev-p replied:
added the check to `system_keyspace::update_peer_info` that `host_id` and `inet_address` are set.

kbr-scylla commented on db/system_keyspace.hh:
```diff
@@ -128,6 +128,7 @@ class system_keyspace : public seastar::peering_sharded_service<system_keyspace>
     static schema_ptr large_cells();
     static schema_ptr scylla_local();
     future<> force_blocking_flush(sstring cfname);
+    future<> ensure_peers_table_fixup();
```
Name is vague. I suggest to either use a more descriptive name (`remove_old_ips_from_peers_table`? but it doesn't cover the host ID part -- which I'm not a fan of though, comment above) or a comment (preferably both)

gusev-p replied:
renamed the functions, added a comment

gusev-p commented:
new version:
  * renamed `ensure_peers_table_fixup -> peers_table_read_fixup`, added a comment

scylladb-promoter commented:
#### :green_circle: CI State: SUCCESS

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/6447/)
:white_check_mark: - [Container Test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/6447/)
:white_check_mark: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/7694/)
:white_check_mark: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/6447/)
#### Build Details:
- Duration: 1 hr 47 min
- Builder: spider8.cloudius-systems.com

gusev-p commented:
new version:
  * rebased on master
  * added the check that host_id and inet_address are set in `system_keyspace::update_peer_info`
  * fixed a bug in `storage_service::on_change` - [we should grab the host_id from the gossiper](https://github.com/scylladb/scylladb/pull/17162/commits/fdc76d3b0a6fce8085dbb84c1f65d5a0a0971175), not from the set of changed app states.

kbr-scylla commented:
@gusev-p please verify whether this
https://github.com/scylladb/scylladb/issues/16691
is fixed by this PR

scylladb-promoter commented:
#### :green_circle: CI State: SUCCESS

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/6546/)
:white_check_mark: - [Container Test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/6546/)
:white_check_mark: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/7785/)
:white_check_mark: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/6546/)
#### Build Details:
- Duration: 2 hr 21 min
- Builder: spider1.cloudius-systems.com

gusev-p commented:
new version:
  * added `fixes scylladb/scylladb#16691` to the commit message

scylladb-promoter commented:
#### :green_circle: CI State: SUCCESS

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/6553/)
:white_check_mark: - [Container Test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/6553/)
:white_check_mark: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/7793/)
:white_check_mark: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/6553/)
#### Build Details:
- Duration: 2 hr 20 min
- Builder: spider6.cloudius-systems.com

kbr-scylla commented on db/system_keyspace.cc:
```diff
@@ -1463,7 +1463,47 @@ struct local_cache {
 };
 
+future<> system_keyspace::peers_table_read_fixup() {
+    static bool done = false;
```
`thread_local` or (better) we should assert that this is called on shard 0 only.

gusev-p replied:
done

kbr-scylla commented on db/system_keyspace.cc:
```diff
@@ -1463,7 +1463,47 @@ struct local_cache {
+    for (const auto& row : *cql_result) {
+        const auto peer = row.get_as<net::inet_address>("peer");
+        if (!row.has("host_id")) {
+            slogger.warn("Peer {} has no host_id in system.{}, the record is broken, removing it",
```
I'll merge, but this should be reported as an ERROR -- it most likely indicates a bug.
Please send a follow-up to address this

gusev-p replied:
done for both problems - missing `host_id` and duplicate `IP`-s.

kbr-scylla commented on test/topology/test_change_ip.py:
```diff
@@ -11,47 +11,69 @@
+        await manager.server_stop(s_0)
+        await manager.server_start(s_0)
+    host0 = (await wait_for_cql_and_get_hosts(cql, [servers[0]], time.time() + 60))[0]
+    await wait_for_peers(host0, {s1_new_ip, s2_old_ip})
```
Previously (the inlined version of) `wait_for_peers` was called after reconnecting driver, which served as a workaround for https://github.com/scylladb/python-driver/issues/295 among other things.

Now you're calling it after a server restarts + wait_for_cql_and_get_hosts --- so the test might now likely hit https://github.com/scylladb/python-driver/issues/295, making it flaky.

I'll merge, the flakiness should be rare, but please send a follow-up to address this (we need a driver reconnect after server restart before doing queries)

gusev-p replied:
done

kbr-scylla commented:
Unfortunately, there's a conflict with current `next`

gusev-p commented:
new version:
  * rebased on master;
  * fixed the review comments;
  * [two](https://github.com/scylladb/scylladb/pull/17162/commits/7559010441f72c1d84dc2909119d95ef26935507) [commits](https://github.com/scylladb/scylladb/pull/17162/commits/29346ab7b1b3629b43aced0b4a696444c37a40d8) were added to fix https://github.com/scylladb/scylladb/issues/17199;
  * `test_change_ip` refactored - added more checks for the problems fixed.

@kbr-scylla

scylladb-promoter commented:
#### :green_circle: CI State: SUCCESS

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/6606/)
:white_check_mark: - [Container Test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/6606/)
:white_check_mark: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/7846/)
:white_check_mark: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/6606/)
#### Build Details:
- Duration: 2 hr 8 min
- Builder: spider4.cloudius-systems.com

kbr-scylla commented:
I don't like it that the scope of the PR suddenly increased, it puts large overhead on the reviewers to understand what changed and what didn't from the previous version, incremental development is easier. Basically a second PR was added to the original one...

kbr-scylla commented:
The cover letter is exactly the same as before, which makes things even more confusing.

kbr-scylla commented on main.cc:
```diff
@@ -1793,8 +1793,10 @@ To start the scylla server proper, simply invoke as: scylla server (or just scyl
             // Set up group0 service earlier since it is needed by group0 setup just below
             ss.local().set_group0(group0_service, raft_topology_change_enabled);
```
> Also we remove the load_initial_raft_address_map function from raft_group0 as it seems redundant. When start_server_for_group0 is called from setup_group0_if_exist gossiper is not initialized yet, so this call has no effect.

But `start_server_for_group0` is also called from the other `setup_group0` -> `join_group0`, that's where it call is needed i.e. when we join the cluster. Not when we restart.

"seems redundant" is not a good enough reason to remove something, there was a reason it was added. The reason is that we might miss all gossiper notifications about endpoints that are delivered by gossiper before we subscribe to it.

kbr-scylla replied:
> I should have given a more detailed reasoning in the commit message, updated it.

The commit message looks the same to me as it was.

I would explain why it's also no longer necessary on the other path where we start server for group -- we subscribe to gossiper before starting gossiping, so we won't miss any gossiper notifications.

gusev-p replied:
> The commit message looks the same to me as it was.

Idk what's happening, sometimes git just refuses to do what I want. I clearly remember how I updated the commit message.

We do this again.

kbr-scylla commented on service/raft/raft_group0.cc:
```diff
@@ -748,25 +742,6 @@ future<> raft_group0::setup_group0(
-        // can safely overwrite existing entries since are loading
-        // them directly from gossiper app state - which is most
-        // recent.
-        _raft_gr.address_map().add_or_update_entry(raft::server_id{server_id}, ip_addr);
```
This should probably provide the generation number though

kbr-scylla commented on service/storage_service.cc:
```diff
@@ -704,9 +704,16 @@ class storage_service::raft_ip_address_updater: public gms::i_endpoint_state_cha
+        if (prev_ip == endpoint) {
+            co_return;
+        }
+        if (_address_map.find(id) == prev_ip) {
```
Comment should explain the reasoning behind this branch

kbr-scylla commented on service/storage_service.cc:
```diff
@@ -728,9 +733,13 @@ class storage_service::raft_ip_address_updater: public gms::i_endpoint_state_cha
-                        return _ss.sync_raft_topology_nodes(std::move(t), hid, prev_ip, {});
-                    }).finally([g = std::move(g)]{});
+                    return utils::get_local_injector().inject("ip-change-raft-sync-delay", std::chrono::milliseconds(500))
+                        .then([this, hid, prev_ip]() {
```
I see we're piling more and more `then`s, perhaps it's time to coroutinize this code?

kbr-scylla commented on test/topology/test_change_ip.py:
```diff
@@ -8,56 +8,118 @@
+    s2_new_ip = await manager.server_change_ip(servers[2].server_id)
+    logger.info(f"s1 new ip {s1_new_ip}, s2 new ip {s2_new_ip}")
+
+    async def reconnect_driver():
```
You can import from test.topology.util

kbr-scylla commented on test/topology/test_change_ip.py:
```diff
@@ -8,56 +8,118 @@
 
 @pytest.mark.asyncio
-async def test_change_two(manager, random_tables):
+async def test_change_two(manager, random_tables, mode):
```
Please execute the test 100 times in debug mode to eliminate potential flakiness

gusev-p replied:
got two failures out of 100, will try to investigate tomorrow

gusev-p replied:
fixed

kbr-scylla commented on test/topology/test_change_ip.py:
```diff
@@ -8,56 +8,118 @@
+        await manager.driver_connect()
+
+    async def wait_proper_ips(alive_servers: list[ServerInfo]):
+        hosts = {h.address: h for h in await wait_for_cql_and_get_hosts(manager.cql, alive_servers, time.time() + 60)}
```
manager.get_cql() to avoid type error

kbr-scylla commented on service/storage_service.cc:
```diff
@@ -431,6 +431,11 @@ future<> storage_service::sync_raft_topology_nodes(mutable_token_metadata_ptr tm
             }
 
             if (prev_ip) {
```
>   * We inject failure before old IP is removed and verify  that after restart the node sees the proper peers - the  new IP for node2 and old IP for node2, which is not restarted  yet.

I think it was supposed to be new IP for node1

gusev-p commented:
> I don't like it that the scope of the PR suddenly increased, it puts large overhead on the reviewers to understand what changed and what didn't from the previous version, incremental development is easier. Basically a second PR was added to the original one...

The changes are tightly coupled, especially in `test_change_ip.py`. With two PRs there would be many conflicts and dependencies. And I believe it's also easier and faster to review - one test_change_ip refactor instead of many changes scattered in different commits/PRs.

gusev-p commented:
> The cover letter is exactly the same as before, which makes things even more confusing.

updated

gusev-p commented:
new version:
  * small review fixes
  * coroutinized [raft_ip_address_updater::on_endpoint_change](https://github.com/scylladb/scylladb/pull/17162/commits/3f18f466539931cef0ae4a7b5e9f2d7c3961a172)

[all changes](https://github.com/scylladb/scylladb/compare/57705690e796a7c05ff04312e91f7227b03bafac..c8c9f485cc86f51fdf9e2defe06dc69b9e71510b)

kbr-scylla commented on service/raft/raft_group0.cc:
```diff
@@ -386,12 +386,6 @@ future<> raft_group0::start_server_for_group0(raft::group_id group0_id, service:
-    // At this time the group registry is already up and running,
-    // so the address map is getting all the notifications from
-    // the gossiper. By reading the application state *after* subscribing to new gossip events,
-    // we ensure we haven't missed any IP update in the map.
```
> I don't see how we can miss anything with the current code - init_address_map is called before gossiper is started, it does both - adds the stored mappings and subscribes to gossiper.

Ok, if we now indeed subscribe to gossiper before starting it, it should work. I think previously the order was reversed, so we could have missed some updates that were delivered before we subscribed, and we needed to fill that missing state by this scan.

kbr-scylla commented on test/topology/test_change_ip.py:
```diff
@@ -8,56 +8,118 @@
+        await s0_logs.wait_for('crash-before-prev-ip-removed hit, killing the node')
+        await manager.server_stop(servers[0].server_id)
+        await manager.server_start(servers[0].server_id)
+        await manager.api.enable_injection(servers[0].ip_addr, 'ip-change-raft-sync-delay', one_shot=False)
```
Depends on which notification this is supposed to catch -- if it's supposed to observe `servers[1]` or itself on restart, that notification may happen before you inject, so the injection will not work. If you're injecting this to catch the `servers[2]` start below, then it's fine.

(BTW. comments would be helpful, why we're delaying these notifications and which notifications we want to delay)

gusev-p replied:
> (BTW. comments would be helpful, why we're delaying these notifications and which notifications we want to delay)

This is mentioned in the commit message.
> Note that we need to inject an additional delay ip-change-raft-sync-delay
> before old IP is removed. Otherwise the problem stop reproducing - other
> nodes remove the old IP before it's send back to the just restarted node.

> Depends on which notification this is supposed to catch -- if it's supposed to observe servers[1] or itself on restart, that notification may happen before you inject, so the injection will not work. (And then you should use the startup injections.) If you're injecting this to catch the servers[2] start below, then it's fine.

Injection into `servers[0]` should catch `server[1]` and `server[2]` restart. The `server[0]` itself restarts `if mode != 'release'`, so we need to inject `ip-change-raft-sync-delay` again after restart.

kbr-scylla replied:
> This is mentioned in the commit message.

I meant comments in the test code. The commit message is much harder to access (you need to `git blame` etc.)

kbr-scylla commented on service/storage_service.cc:
This problem was solved in existing code by turning the captures into arguments and then binding them with `bind_front`. See e.g. `raft_group0::discover_group0`
```cpp
    }).finally(std::bind_front([] (raft_group0& self, persistent_discovery& p_discovery) -> future<> {
        co_await p_discovery.stop();
        self._group0 = std::monostate{};
    }, std::ref(*this), std::ref(p_discovery)));
```

kbr-scylla replied:
Also why leave this single `then`? It could be put into the coroutine too.

scylladb-promoter commented:
#### :green_circle: CI State: SUCCESS

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/6634/)
:white_check_mark: - [Container Test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/6634/)
:white_check_mark: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/6634/)
:white_check_mark: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/7866/)
#### Build Details:
- Duration: 1 hr 53 min
- Builder: spider1.cloudius-systems.com

gusev-p commented on service/storage_service.cc:
done

gusev-p commented:
new version:
  * rebased on master;
  * make `on_endpoint_change` even more coroutinized;
  * restored the lost commit message part about why `load_initial_raft_address_map` was removed;
  * use `unique_ptr` in `ensure_alive`.

scylladb-promoter commented:
#### :green_circle: CI State: SUCCESS

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/6638/)
:white_check_mark: - [Container Test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/6638/)
:white_check_mark: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/7871/)
:white_check_mark: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/6638/)
#### Build Details:
- Duration: 2 hr 1 min
- Builder: spider5.cloudius-systems.com

bhalevy commented on db/system_keyspace.cc:
```diff
@@ -1735,114 +1773,98 @@ static std::vector<cdc::generation_id_v2> decode_cdc_generations_ids(const set_t
-            }
+    co_await peers_table_read_fixup();
+
+    const sstring req = format("SELECT peer, host_id, tokens FROM system.{}", PEERS);
```
nit: host_id is unused now

bhalevy commented on db/system_keyspace.cc:
```diff
@@ -1735,114 +1773,98 @@ static std::vector<cdc::generation_id_v2> decode_cdc_generations_ids(const set_t
-    auto res = co_await execute_cql(format("SELECT peer, host_id, tokens FROM system.{}", PEERS));
+    co_await peers_table_read_fixup();
+
+    const auto res = co_await execute_cql(format("SELECT peer, host_id, tokens FROM system.{}", PEERS));
```
nit: host_id is unused now

gusev-p commented:
new version:
  * dropped background fibers from `raft_ip_address_updater::on_endpoint_change`, use `gms::permit_id` instead
  * make the `./test.py topology/test_change_ip --mode debug --save-log-on-success --repeat 100` pass
    * fixed the bug in old IP removal when a node changes IP -- passing `prev_ip` to `sync_raft_topology_nodes` doesn't work, since the node may be not in a normal state at the moment, we won't use `prev_ip` neither in the current `sync_raft_topology_nodes` call not in the subsequent `topology_state_load` where it will eventually become normal. Fixed it by constructing `host_id -> ip` map inside `sync_raft_topology_nodes` and consulting it every time we update a normal node.

scylladb-promoter commented:
#### :red_circle: CI State: FAILURE

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/6672/)
:white_check_mark: - [Container Test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/6672/)
:x: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/6672/)
:white_check_mark: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/7903/)
#### Failed Tests (39/24652):
- [test_no_cleanup_when_unnecessary](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/6672/testReport/junit/%28root%29/test_automatic_cleanup/Tests___Unit_Tests___test_no_cleanup_when_unnecessary) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_no_cleanup_when_unnecessary)
- [test_no_cleanup_when_unnecessary](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/6672/testReport/junit/%28root%29/test_automatic_cleanup/Tests___Unit_Tests___test_no_cleanup_when_unnecessary_2) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_no_cleanup_when_unnecessary)
- [test_snapshot](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/6672/testReport/junit/%28root%29/test_snapshot/Tests___Unit_Tests___test_snapshot) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_snapshot)
- [test_snapshot](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/6672/testReport/junit/%28root%29/test_snapshot/Tests___Unit_Tests___test_snapshot_2) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_snapshot)
- [test_topology_streaming_failure](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/6672/testReport/junit/%28root%29/test_topology_failure_recovery/Tests___Unit_Tests___test_topology_streaming_failure) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_topology_streaming_failure)
- [test_topology_streaming_failure](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/6672/testReport/junit/%28root%29/test_topology_failure_recovery/Tests___Unit_Tests___test_topology_streaming_failure_2) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_topology_streaming_failure)
- [test_remove_node_add_column](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/6672/testReport/junit/%28root%29/test_topology_remove_decom/Tests___Unit_Tests___test_remove_node_add_column) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_remove_node_add_column)
- [test_decommission_node_add_column](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/6672/testReport/junit/%28root%29/test_topology_remove_decom/Tests___Unit_Tests___test_decommission_node_add_column) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_decommission_node_add_column)
- [test_decommission_node_add_column](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/6672/testReport/junit/%28root%29/test_topology_remove_decom/Tests___Unit_Tests___test_decommission_node_add_column_2) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_decommission_node_add_column)
- [test_remove_node_add_column](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/6672/testReport/junit/%28root%29/test_topology_remove_decom/Tests___Unit_Tests___test_remove_node_add_column_2) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_remove_node_add_column)
#### Build Details:
- Duration: 3 hr 5 min
- Builder: spider6.cloudius-systems.com

kbr-scylla commented:
@gusev-p I looked at one of these failures, it timed out on server_add, and e.g.
https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/6672/artifact/testlog/x86_64/debug/scylla-1303.log
```
INFO  2024-02-14 22:14:30,922 [shard 0:strm] query_processor - Column definitions for system_distributed.cdc_generation_timestamps changed, invalidating related prepared statements
INFO  2024-02-14 22:14:30,925 [shard 0:strm] schema_tables - Schema version changed to a31f558e-cb75-11ee-0e0a-4a3ffd80a9f8
INFO  2024-02-14 22:14:30,926 [shard 0:strm] messaging_service - Dropping connection to 127.35.2.8 because it was created without topology information
INFO  2024-02-14 22:14:30,926 [shard 0:strm] messaging_service - Dropping connection to 127.35.2.8 because it was created without topology information
INFO  2024-02-14 22:14:30,929 [shard 1:strm] messaging_service - Dropping connection to 127.35.2.11 because it was created without topology information
INFO  2024-02-14 22:14:30,930 [shard 0:strm] messaging_service - Dropping connection to 127.35.2.16 because it was created without topology information
INFO  2024-02-14 22:15:02,454 [shard 0:main] raft_group_registry - marking Raft server da6421f5-1938-4eb5-8d3b-583d42d00214 as dead for raft groups
INFO  2024-02-14 22:15:02,476 [shard 0:main] raft_group_registry - marking Raft server 5cff46f7-d35a-4502-ab56-97724a2bb50b as dead for raft groups
INFO  2024-02-14 22:15:02,477 [shard 0:main] raft_group_registry - marking Raft server becf031c-b903-46df-869f-9cafb4a643d7 as dead for raft groups
WARN  2024-02-14 22:15:03,273 [shard 0:strm] gossip - failure_detector_loop: Send echo to node 127.35.2.8, status = failed: seastar::rpc::closed_error (connection is closed)
WARN  2024-02-14 22:15:05,275 [shard 0:strm] gossip - failure_detector_loop: Send echo to node 127.35.2.8, status = failed: seastar::rpc::closed_error (connection is closed)
WARN  2024-02-14 22:15:07,277 [shard 0:strm] gossip - failure_detector_loop: Send echo to node 127.35.2.8, status = failed: seastar::rpc::closed_error (connection is closed)
WARN  2024-02-14 22:15:09,278 [shard 0:strm] gossip - failure_detector_loop: Send echo to node 127.35.2.8, status = failed: seastar::rpc::closed_error (connection is closed)
WARN  2024-02-14 22:15:11,280 [shard 0:strm] gossip - failure_detector_loop: Send echo to node 127.35.2.8, status = failed: seastar::rpc::closed_error (connection is closed)
WARN  2024-02-14 22:15:13,282 [shard 0:strm] gossip - failure_detector_loop: Send echo to node 127.35.2.8, status = failed: seastar::rpc::closed_error (connection is closed)
WARN  2024-02-14 22:15:15,283 [shard 0:strm] gossip - failure_detector_loop: Send echo to node 127.35.2.8, status = failed: seastar::rpc::closed_error (connection is closed)
WARN  2024-02-14 22:15:17,285 [shard 0:strm] gossip - failure_detector_loop: Send echo to node 127.35.2.8, status = failed: seastar::rpc::closed_error (connection is closed)
WARN  2024-02-14 22:15:19,287 [shard 0:strm] gossip - failure_detector_loop: Send echo to node 127.35.2.8, status = failed: seastar::rpc::closed_error (connection is closed)
WARN  2024-02-14 22:15:21,289 [shard 0:strm] gossip - failure_detector_loop: Send echo to node 127.35.2.8, status = failed: seastar::rpc::closed_error (connection is closed)
WARN  2024-02-14 22:15:23,291 [shard 0:strm] gossip - failure_detector_loop: Send echo to node 127.35.2.8, status = failed: seastar::rpc::closed_error (connection is closed)
INFO  2024-02-14 22:15:23,291 [shard 0:strm] gossip - failure_detector_loop: Mark node 127.35.2.8 as DOWN
INFO  2024-02-14 22:15:23,292 [shard 0:strm] gossip - InetAddress 127.35.2.8 is now DOWN, status = NORMAL
ERROR 2024-02-14 22:15:30,629 [shard 0: gms] gossip - gms/gossiper.cc(1689:28) `future<> gms::gossiper::real_mark_alive(inet_address)`: waiting for endpoint lock (ep=127.35.2.16) took more than 1m, signifying possible deadlock; holders: 1, first holder: gms/gossiper.cc(569:28) `future<> gms::gossiper::do_apply_state_locally(gms::inet_address, endpoint_state, bool)`, last holder (might not be current): gms/gossiper.cc(569:28) `future<> gms::gossiper::do_apply_state_locally(gms::inet_address, endpoint_state, bool)`
ERROR 2024-02-14 22:15:30,629 [shard 0: gms] gossip - gms/gossiper.cc(1689:28) `future<> gms::gossiper::real_mark_alive(inet_address)`: waiting for endpoint lock (ep=127.35.2.11) took more than 1m, signifying possible deadlock; holders: 1, first holder: gms/gossiper.cc(569:28) `future<> gms::gossiper::do_apply_state_locally(gms::inet_address, endpoint_state, bool)`, last holder (might not be current): gms/gossiper.cc(569:28) `future<> gms::gossiper::do_apply_state_locally(gms::inet_address, endpoint_state, bool)`
ERROR 2024-02-14 22:15:30,764 [shard 0: gms] gossip - gms/gossiper.cc(569:28) `future<> gms::gossiper::do_apply_state_locally(gms::inet_address, endpoint_state, bool)`: waiting for endpoint lock (ep=127.35.2.16) took more than 1m, signifying possible deadlock; holders: 1, first holder: gms/gossiper.cc(569:28) `future<> gms::gossiper::do_apply_state_locally(gms::inet_address, endpoint_state, bool)`, last holder (might not be current): gms/gossiper.cc(569:28) `future<> gms::gossiper::do_apply_state_locally(gms::inet_address, endpoint_state, bool)`
ERROR 2024-02-14 22:15:30,788 [shard 0: gms] gossip - gms/gossiper.cc(569:28) `future<> gms::gossiper::do_apply_state_locally(gms::inet_address, endpoint_state, bool)`: waiting for endpoint lock (ep=127.35.2.11) took more than 1m, signifying possible deadlock; holders: 1, first holder: gms/gossiper.cc(569:28) `future<> gms::gossiper::do_apply_state_locally(gms::inet_address, endpoint_state, bool)`, last holder (might not be current): gms/gossiper.cc(569:28) `future<> gms::gossiper::do_apply_state_locally(gms::inet_address, endpoint_state, bool)`
ERROR 2024-02-14 22:15:30,934 [shard 0:strm] gossip - service/storage_service.cc(670:32) `future<> service::storage_service::topology_state_load()`: waiting for endpoint lock (ep=127.35.2.11) took more than 1m, signifying possible deadlock; holders: 1, first holder: gms/gossiper.cc(569:28) `future<> gms::gossiper::do_apply_state_locally(gms::inet_address, endpoint_state, bool)`, last holder (might not be current): gms/gossiper.cc(569:28) `future<> gms::gossiper::do_apply_state_locally(gms::inet_address, endpoint_state, bool)`
ERROR 2024-02-14 22:15:31,261 [shard 0: gms] gossip - gms/gossiper.cc(736:32) `future<> gms::gossiper::do_status_check()`: waiting for endpoint lock (ep=127.35.2.16) took more than 1m, signifying possible deadlock; holders: 1, first holder: gms/gossiper.cc(569:28) `future<> gms::gossiper::do_apply_state_locally(gms::inet_address, endpoint_state, bool)`, last holder (might not be current): gms/gossiper.cc(569:28) `future<> gms::gossiper::do_apply_state_locally(gms::inet_address, endpoint_state, bool)`
```

and so on, gossiper is deadlocked

kbr-scylla commented:
Could be from
> dropped background fibers from raft_ip_address_updater::on_endpoint_change, use gms::permit_id instead

gusev-p commented:
> @gusev-p I looked at one of these failures, it timed out on server_add, and e.g. https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/6672/artifact/testlog/x86_64/debug/scylla-1303.log
> 
> ```
> INFO  2024-02-14 22:14:30,922 [shard 0:strm] query_processor - Column definitions for system_distributed.cdc_generation_timestamps changed, invalidating related prepared statements
> INFO  2024-02-14 22:14:30,925 [shard 0:strm] schema_tables - Schema version changed to a31f558e-cb75-11ee-0e0a-4a3ffd80a9f8
> INFO  2024-02-14 22:14:30,926 [shard 0:strm] messaging_service - Dropping connection to 127.35.2.8 because it was created without topology information
> INFO  2024-02-14 22:14:30,926 [shard 0:strm] messaging_service - Dropping connection to 127.35.2.8 because it was created without topology information
> INFO  2024-02-14 22:14:30,929 [shard 1:strm] messaging_service - Dropping connection to 127.35.2.11 because it was created without topology information
> INFO  2024-02-14 22:14:30,930 [shard 0:strm] messaging_service - Dropping connection to 127.35.2.16 because it was created without topology information
> INFO  2024-02-14 22:15:02,454 [shard 0:main] raft_group_registry - marking Raft server da6421f5-1938-4eb5-8d3b-583d42d00214 as dead for raft groups
> INFO  2024-02-14 22:15:02,476 [shard 0:main] raft_group_registry - marking Raft server 5cff46f7-d35a-4502-ab56-97724a2bb50b as dead for raft groups
> INFO  2024-02-14 22:15:02,477 [shard 0:main] raft_group_registry - marking Raft server becf031c-b903-46df-869f-9cafb4a643d7 as dead for raft groups
> WARN  2024-02-14 22:15:03,273 [shard 0:strm] gossip - failure_detector_loop: Send echo to node 127.35.2.8, status = failed: seastar::rpc::closed_error (connection is closed)
> WARN  2024-02-14 22:15:05,275 [shard 0:strm] gossip - failure_detector_loop: Send echo to node 127.35.2.8, status = failed: seastar::rpc::closed_error (connection is closed)
> WARN  2024-02-14 22:15:07,277 [shard 0:strm] gossip - failure_detector_loop: Send echo to node 127.35.2.8, status = failed: seastar::rpc::closed_error (connection is closed)
> WARN  2024-02-14 22:15:09,278 [shard 0:strm] gossip - failure_detector_loop: Send echo to node 127.35.2.8, status = failed: seastar::rpc::closed_error (connection is closed)
> WARN  2024-02-14 22:15:11,280 [shard 0:strm] gossip - failure_detector_loop: Send echo to node 127.35.2.8, status = failed: seastar::rpc::closed_error (connection is closed)
> WARN  2024-02-14 22:15:13,282 [shard 0:strm] gossip - failure_detector_loop: Send echo to node 127.35.2.8, status = failed: seastar::rpc::closed_error (connection is closed)
> WARN  2024-02-14 22:15:15,283 [shard 0:strm] gossip - failure_detector_loop: Send echo to node 127.35.2.8, status = failed: seastar::rpc::closed_error (connection is closed)
> WARN  2024-02-14 22:15:17,285 [shard 0:strm] gossip - failure_detector_loop: Send echo to node 127.35.2.8, status = failed: seastar::rpc::closed_error (connection is closed)
> WARN  2024-02-14 22:15:19,287 [shard 0:strm] gossip - failure_detector_loop: Send echo to node 127.35.2.8, status = failed: seastar::rpc::closed_error (connection is closed)
> WARN  2024-02-14 22:15:21,289 [shard 0:strm] gossip - failure_detector_loop: Send echo to node 127.35.2.8, status = failed: seastar::rpc::closed_error (connection is closed)
> WARN  2024-02-14 22:15:23,291 [shard 0:strm] gossip - failure_detector_loop: Send echo to node 127.35.2.8, status = failed: seastar::rpc::closed_error (connection is closed)
> INFO  2024-02-14 22:15:23,291 [shard 0:strm] gossip - failure_detector_loop: Mark node 127.35.2.8 as DOWN
> INFO  2024-02-14 22:15:23,292 [shard 0:strm] gossip - InetAddress 127.35.2.8 is now DOWN, status = NORMAL
> ERROR 2024-02-14 22:15:30,629 [shard 0: gms] gossip - gms/gossiper.cc(1689:28) `future<> gms::gossiper::real_mark_alive(inet_address)`: waiting for endpoint lock (ep=127.35.2.16) took more than 1m, signifying possible deadlock; holders: 1, first holder: gms/gossiper.cc(569:28) `future<> gms::gossiper::do_apply_state_locally(gms::inet_address, endpoint_state, bool)`, last holder (might not be current): gms/gossiper.cc(569:28) `future<> gms::gossiper::do_apply_state_locally(gms::inet_address, endpoint_state, bool)`
> ERROR 2024-02-14 22:15:30,629 [shard 0: gms] gossip - gms/gossiper.cc(1689:28) `future<> gms::gossiper::real_mark_alive(inet_address)`: waiting for endpoint lock (ep=127.35.2.11) took more than 1m, signifying possible deadlock; holders: 1, first holder: gms/gossiper.cc(569:28) `future<> gms::gossiper::do_apply_state_locally(gms::inet_address, endpoint_state, bool)`, last holder (might not be current): gms/gossiper.cc(569:28) `future<> gms::gossiper::do_apply_state_locally(gms::inet_address, endpoint_state, bool)`
> ERROR 2024-02-14 22:15:30,764 [shard 0: gms] gossip - gms/gossiper.cc(569:28) `future<> gms::gossiper::do_apply_state_locally(gms::inet_address, endpoint_state, bool)`: waiting for endpoint lock (ep=127.35.2.16) took more than 1m, signifying possible deadlock; holders: 1, first holder: gms/gossiper.cc(569:28) `future<> gms::gossiper::do_apply_state_locally(gms::inet_address, endpoint_state, bool)`, last holder (might not be current): gms/gossiper.cc(569:28) `future<> gms::gossiper::do_apply_state_locally(gms::inet_address, endpoint_state, bool)`
> ERROR 2024-02-14 22:15:30,788 [shard 0: gms] gossip - gms/gossiper.cc(569:28) `future<> gms::gossiper::do_apply_state_locally(gms::inet_address, endpoint_state, bool)`: waiting for endpoint lock (ep=127.35.2.11) took more than 1m, signifying possible deadlock; holders: 1, first holder: gms/gossiper.cc(569:28) `future<> gms::gossiper::do_apply_state_locally(gms::inet_address, endpoint_state, bool)`, last holder (might not be current): gms/gossiper.cc(569:28) `future<> gms::gossiper::do_apply_state_locally(gms::inet_address, endpoint_state, bool)`
> ERROR 2024-02-14 22:15:30,934 [shard 0:strm] gossip - service/storage_service.cc(670:32) `future<> service::storage_service::topology_state_load()`: waiting for endpoint lock (ep=127.35.2.11) took more than 1m, signifying possible deadlock; holders: 1, first holder: gms/gossiper.cc(569:28) `future<> gms::gossiper::do_apply_state_locally(gms::inet_address, endpoint_state, bool)`, last holder (might not be current): gms/gossiper.cc(569:28) `future<> gms::gossiper::do_apply_state_locally(gms::inet_address, endpoint_state, bool)`
> ERROR 2024-02-14 22:15:31,261 [shard 0: gms] gossip - gms/gossiper.cc(736:32) `future<> gms::gossiper::do_status_check()`: waiting for endpoint lock (ep=127.35.2.16) took more than 1m, signifying possible deadlock; holders: 1, first holder: gms/gossiper.cc(569:28) `future<> gms::gossiper::do_apply_state_locally(gms::inet_address, endpoint_state, bool)`, last holder (might not be current): gms/gossiper.cc(569:28) `future<> gms::gossiper::do_apply_state_locally(gms::inet_address, endpoint_state, bool)`
> ```
> 
> and so on, gossiper is deadlocked

yes, same problem fixed by `ensure_alive`, gossiper permits don't help here, reverting back

gusev-p commented:
new version:
  * fix tests, revert the stupid attempt at solving gossiper deadlock with `permit_id`-s, it doesn't work

scylladb-promoter commented:
#### :green_circle: CI State: SUCCESS

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/6690/)
:white_check_mark: - [Container Test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/6690/)
:white_check_mark: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/7917/)
:white_check_mark: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/6690/)
#### Build Details:
- Duration: 1 hr 55 min
- Builder: spider2.cloudius-systems.com

kbr-scylla commented on service/storage_service.cc:
```diff
@@ -370,16 +370,37 @@ future<> storage_service::sync_raft_topology_nodes(mutable_token_metadata_ptr tm
+        sys_ks_futures.push_back(_sys_ks.local().remove_endpoint(ip));
+
+        if (_gossiper.get_endpoint_state_ptr(ip) && !get_used_ips().contains(ip)) {
+            co_await _gossiper.force_remove_endpoint(ip, gms::null_permit_id);
```
Haven't noticed this earlier, but is it OK to do it before / concurrently with `_sys_ks.local().remove_endpoint(ip));`?

gusev-p replied:
> Hmm, I think this function is bugged currently, and gossiper updates need to happen after corresponding system keyspace updates.

why?

kbr-scylla replied:
I'm not sure about remove_ip, the whole thing is hard to reason about..

but for example here
```cpp
            sys_ks_futures.push_back(_sys_ks.local().update_peer_info(*ip, info));
            if (!prev_normal.contains(id)) {
                co_await notify_joined(*ip);
            }
```

isn't it important that we put entry to peers before we send notifications? cc @gleb-cloudius 

Driver could receive such notification, then refresh peers but observe old state.

In general outside observes may see things happening out of order, and I'm worried it could cause hard to reproduce and debug bugs.


Or here
```cpp
            sys_ks_futures.push_back(_sys_ks.local().update_tokens(rs.ring.value().tokens));
            co_await _gossiper.add_local_application_state({
                { gms::application_state::TOKENS, gms::versioned_value::tokens(rs.ring.value().tokens) },
                { gms::application_state::CDC_GENERATION_ID, gms::versioned_value::cdc_generation_id(_topology_state_machine._topology.current_cdc_generation_id) },
                { gms::application_state::STATUS, gms::versioned_value::normal(rs.ring.value().tokens) }
            });
```

is it fine that other nodes might observe our gossip status update before we persist it locally? I think it's just asking for trouble...

kbr-scylla replied:
Moving the concurrency "outside" these lambdas would simplify reasoning a lot, put some order to things, while keeping the optimization intact. (There would still be concurrency between updates for different peers, which was the goal.)

gusev-p replied:
> The concurrency of system keyspace updates can be achieved by not co_awaiting on the process_normal_node etc. calls but putting them into a vector of futures. But inside process_normal_node etc., we should be doing operations in order.

It can't, this'll entail concurrent `token_metadata` mutations, e.g. multiple fibers mutating `_token_to_endpoint_map` through `update_normal_tokens`.

kbr-scylla replied:
Anyway, it's out of scope of this PR.

kbr-scylla commented on service/storage_service.cc:
```diff
@@ -425,6 +451,10 @@ future<> storage_service::sync_raft_topology_nodes(mutable_token_metadata_ptr tm
                 co_await notify_joined(*ip);
             }
+
+            if (const auto it = host_id_to_ip_map.find(host_id); it != host_id_to_ip_map.end() && it->second != *ip) {
```
I don't see any code that would add / update new mapping in this map, if one appears

kbr-scylla replied:
So it looks like we'll purge the old IP only in very specific scenario, the first time we change a node's IP... Then when it's changed again into a third IP, we won't find the second, so we won't purge it

gusev-p replied:
`host_id_to_ip_map` is local to the `sync_raft_topology_nodes` invocation, each node is processed only once

gusev-p replied:
it's a local multishard read from memory (it should reside in row_cache), how  can it be more expensive than e.g. `o_dsync` write?

kbr-scylla replied:
Anyway the simplest way would be to run the 100 node again and see whether there's any significant regression, maybe my worries are unfounded, @gusev-p could you please check?

kbr-scylla replied:
We can address in follow-up if it turns out it's a problem.

gusev-p replied:
I ran it (still getting 1000-sec timeout):
`INFO  2024-02-15 17:37:43,406 [shard 0:strm] raft_topology - sync_raft_topology_nodes, total us 289570, transition us 0, left us 0, total_update_peer_info_us 1158, total_update_peer_info_count 60, total_normal_us 1838, total_normal_count 61 get_host_id_to_ip_map_us 373 total_sys_ks_us 287646`

kbr-scylla replied:
The previous result was (https://github.com/scylladb/scylladb/issues/17039#issuecomment-1935634888)
```
INFO  2024-02-09 10:58:14,679 [shard 0:strm] raft_topology - sync_raft_topology_nodes, total us 169330, transition us 76, left us 0, total_normal_us 1072, total_normal_count 57
```

so there is some regression, although better what we started with

(or maybe it's variability)

kbr-scylla commented:
Qd
